### PR TITLE
Gracefully stop host if hosted service stops the application during startup

### LINKED ISSUE #51613: Hosted services are started even though the application is instructed to stop.
### Description

As `Microsoft.Extensions.Hosting.Internal.Host` starts, it resolves all registered instances of  `IHostedService` and starts them in sequence. If a hosted service stops the application (using `IHostApplicationLifetime.StopApplication()`), the remaining hosted services are still started (that is `StartAsync` is called), despite the fact that the application is instructed to shut down.

This can results in a combination of the following behaviors:

* If any of the remaining hosted services checks the provided cancellation token and throws an exception, it propagates and results in an unhandled exception unless the user wraps the start call in try/catch. This feels unexpected, as the stopping should be considered graceful.
* If the remaining hosted services does not check the cancellation token, time and resources may be spent starting components that will just be stopped shortly after. This does not scale well if there are "a lot" of registered hosted services that takes "long" time to start.
 
I think it would make sense to check the `IHostApplicationLifetime.ApplicationStopping` cancellation token after each started hosted service and if it is triggered, gracefully exit the application.

One could also argue that `ThrowIfCancellationRequested` should be called on the cancellation token passed to `StartAsync`  after each started service.

Looking forward hearing your thoughts on this!

### PR DESCRIPTION
**Pull request by pardahlman**
This PR closes #51613 by introducing the following behavior when starting `IHostedService` from the `Host`:

1. Before starting each hosted service, check  `IHostApplicationLifetime` to see if the application has been instructed to stop and if so abort the start sequence.
2. If the application is not instructed to stop, check the provided `CancellationToken` to see if the caller has requested cancellation and if so throws `OperationCanceledException`.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs [hunk 1]
```diff
@@ -61,6 +61,15 @@ public async Task StartAsync(CancellationToken cancellationToken = default)
 
             foreach (IHostedService hostedService in _hostedServices)
             {
+                if (_applicationLifetime.ApplicationStopping.IsCancellationRequested)
+                {
+                    var startedServices = _hostedServices.TakeWhile(s => s != hostedService);
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:66]: I'd make some changes to this code:
- Don't call StopAsync from StartAsync, just return early from the loop.
- Change `_hostedServices` to represent the hosted services that successfully started instead of the entire list.
- Store the current list of hosted services in a local and enumerate that in start.

File: src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Internal/HostTests.cs [hunk 2]
```diff
@@ -628,6 +628,57 @@ public async Task HostStopAsyncUsesDefaultTimeoutIfGivenTokenDoesNotFire()
             }
         }
 
+        [Fact]
+        public async Task HostDoesNotStartHostedServiceIfCancellationIsRequested()
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Internal/HostTests.cs:632]: Add 2 tests
1. One that verifies no hosted services start if the host calls `IHostApplicationLifetime.StopApplication` before calling StartAsync.
2. Another that uses IHostApplicationLifetime.StopApplication inside of an IHostedService.StartAsync (that's the scenario you were testing right?)

Comment (pardahlman) [reply src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Internal/HostTests.cs:632]: Thank you for this! 

1. If I understand you correctly and read the code right, you are asking for a test [for this existing line of code](https://github.com/dotnet/runtime/blob/025e7f60cb355f9c44a5ff20784118e785891a4c/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L59) with `_applicationLifetime.ApplicationStopping` being cancelled? Or are you thinking in the lines of calling `IHostApplicationLifetime.StopApplication` as the hosted services [are being resolved from DI](https://github.com/dotnet/runtime/blob/025e7f60cb355f9c44a5ff20784118e785891a4c/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L60)?
2. I believe that this scenario is covered in `HostStopsGracefullyIfHostedServiceStopsTheApplication`, where the first hosted service [calls StopApplication](https://github.com/dotnet/runtime/pull/52517/files#diff-0004f2f896ea10dbe221c137338cf0bf30e116223a38d91911966c58b4d5b5beR673). Let me know if there is something I missed here!

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs (hunk 1 above)
Comment (pardahlman) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs]: I've updated the PR according to the suggestions you made 👍 By doing so, I changed a null check to check if `_hostedServices.Count != 0`, not sure if this check is necessary anymore 🤔 Looking forward hearing your thought on the latest changes!

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs]: Seems like it would be wrong to fire NotifyStarted if we didn't successfully start but I'm not sure. It's weird to return from StartAsync without an error if we haven't successfully started.

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs [hunk 3]
```diff
@@ -21,7 +21,7 @@ internal sealed class Host : IHost, IAsyncDisposable
         private readonly HostOptions _options;
         private readonly IHostEnvironment _hostEnvironment;
         private readonly PhysicalFileProvider _defaultProvider;
-        private IEnumerable<IHostedService> _hostedServices;
+        private readonly ICollection<IHostedService> _hostedServices = new List<IHostedService>();
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:24]:
```suggestion
        private readonly List<IHostedService> _hostedServices = new List<IHostedService>();
```

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs [hunk 4]
```diff
@@ -105,7 +114,7 @@ public async Task StopAsync(CancellationToken cancellationToken = default)
                 _applicationLifetime.StopApplication();
 
                 IList<Exception> exceptions = new List<Exception>();
-                if (_hostedServices != null) // Started?
+                if (_hostedServices.Count != 0) // Started?
```
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:117]:
```suggestion
                if (_hostedServices.Count > 0) // Started?
```

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs (hunk 1 above)
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs]: cc @eerhardt @maryamariyan for opinions

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs [hunk 5]
```diff
@@ -105,7 +114,7 @@ public async Task StopAsync(CancellationToken cancellationToken = default)
                 _applicationLifetime.StopApplication();
 
                 IList<Exception> exceptions = new List<Exception>();
-                if (_hostedServices != null) // Started?
+                if (_hostedServices.Count > 0) // Started?
                 {
                     foreach (IHostedService hostedService in _hostedServices.Reverse())
```
Comment (watfordsuzy) [inline src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:119]: You've hit [List<T>.Reverse](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.reverse?view=net-5.0#System_Collections_Generic_List_1_Reverse), instead of the original LINQ call:
```suggestion
                    foreach (IHostedService hostedService in _hostedServices.AsEnumerable().Reverse())
```

Comment (pardahlman) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:119]: Thanks for solving that mystery 🙏 I believe this was introduced as I accepted [this suggestion](https://github.com/dotnet/runtime/pull/52517/commits/720f78362c8edc5b281fe2eddaeb2f1b0a645b48). Is there an argument for reverting that change instead of applying this fix? Changing back to use `ICollection<IHostedService>` will result in a smaller diff in the MR, but perhaps it is worth applying your fix as it as it is more explicit. Any thought on this?

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:119]: Change the loop to a for loop and enumerate in reverse

Comment (pardahlman) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs]: That's a great suggestion 👌 I've added a new commit to the PR with this change. Let me know if you prefer fixups instead.

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs [hunk 6]
```diff
@@ -57,17 +57,26 @@ public async Task StartAsync(CancellationToken cancellationToken = default)
             await _hostLifetime.WaitForStartAsync(combinedCancellationToken).ConfigureAwait(false);
 
             combinedCancellationToken.ThrowIfCancellationRequested();
-            _hostedServices = Services.GetService<IEnumerable<IHostedService>>();
+            var hostedServices = Services.GetService<IEnumerable<IHostedService>>();
 
-            foreach (IHostedService hostedService in _hostedServices)
+            foreach (IHostedService hostedService in hostedServices)
             {
+                if (_applicationLifetime.ApplicationStopping.IsCancellationRequested)
```
[the remark this answers is not in the document]
Comment (pardahlman) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:64]: Thank you for this feedback 👍 The reasoning for this implementation is to get consistent behavior for calls to `IHostApplicationLifetime.StopApplication()`, despite the state of the host. I would argue that it is expected to get a graceful stop even though the call is made before the host is fully started.

However, I think that if the user provided `CancellationToken` is cancelled, it should throw, since this is expected behavior from other APIs that accept a `CancellationToken`.

So to sum it up: before starting each hosted service, check if the application is instructed to stop, if so, return from the Start method and let the host begin continue to stop. Next, check if the user has requested cancellation, and if so throw exception.

Looking forward hearing your thoughts on this!

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:64]: Did you see @davidfowl's comments here?

https://github.com/dotnet/runtime/pull/52517#discussion_r629436543

> It's weird to return from StartAsync without an error if we haven't successfully started.

[the remark this answers is not in the document]
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs:64]: Maybe it's fine but I haven't thought enough about the implications. Is there anything that would be broken if the started event didn't fire but StartAsync returns successfully?

File: src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs (hunk 1 above)
Comment (pardahlman) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs]: I've been giving this some more thought, and here's my two cents: I think it is correct to return early without calling `NotifyStarted`. The reason why I think this makes sense it that the `IHostApplicationLifetime` would end up in a state where `ApplicationStopping` is triggered, but `ApplicationStarted` is not. To me, that signals that the application is not fully started, but is shutting down - which is also what is happening. I think there is an argument that this is a successful outcome of an graceful shutdown during start.

If, on the other hand, `StartAsync` would throw if the application is stopping before all hosted services are started, that means that there is no way to make a graceful stop when the host is in its start state. Is that desirable?

I appreciate that I do not have the picture here, and I'm looking forward hearing your thoughts on this!

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs]: Yes I can see how this makes sense, my only concern is that we don't have an indication from the start call that it didn't complete. That might be surprising but maybe we can mitigate this with logging.  As a concrete example of what would happen when you use this, [these logs](https://github.com/dotnet/runtime/blob/09ad9e7eb457e5bc911b2292dc557dc02be78a20/src/libraries/Microsoft.Extensions.Hosting/src/Internal/ConsoleLifetime.cs#L69-L71) would never show up. Is that a fine? 

The other thing that could happen is that code would continue executing that is potentially not fully initialized. Code that assumes your hosted services ran and setup some initial state that is then used after the host has started. Maybe the code just needs to be robust against that?

```C#
public class Initializer : IHostedService
{
     private IMyCache _cache;
     public Initializer(IMyCache cache) 
     {
          _cache = cache;
     }
     public Task StartAsync() => _cache["expensiveWork"] = ExpensiveCalculation();
}
```

Then in Main:

```C#
static void Main(string[] args)
{
    using var host = CreateDefaultBuilder(args).Build();
    host.Start();
    var cache = host.Services.GetRequiredService<IMyCache>();
    // Who needs defensive coding
    Use(cache.Cache["expensiveWork"].Value); // This will go boom
}
```

Comment (pardahlman) [reply src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs]: I see what you mean, and understand your concerns 👌 Following the same logical argument as earlier, I think it is correct not to log _Application started. Press Ctrl+C to shut down._, since the application did not transit from starting to started state. I definitely think this can be made clearer by adding logging, as you suggest.

A few thoughts on your cache example:

1. I don't think it is unreasonable to ask the developer to check the state of the host (using `IHostApplicationLifetime`) here. Correct me if I'm wrong, but it is the one source of truth for the host's state. I understand that what I'm suggesting is a changed behavior. It would be interesting to know how common of a scenario it is to have code like that in combination with a call to `StopApplication` made in a hosted service. 
2. If we would throw on `_applicationLifetime.ApplicationStopping` instead of returning, the behavior changes from how it works today, as it is currently possible that all hosted services starts even though the application is instructed to stop.
3. A wild idea, that probably is not feasible:  It is possible to decorate the host's `IServiceProvider` so that if services are resolved from a not fully started host, it logs a warning before resolving the service, or perhaps it should not be possible to resolve services if the start is interrupted? This is probably overkill, and the logging from the incomplete start would probably suffice - just widening the solution space 😆

**Comment by ericstj** [comment]
reviewers, looks like this is pending action on our side cc @davidfowl @eerhardt @maryamariyan

**Comment by eerhardt** [comment]
@pardahlman - I'm still not convinced this is the correct behavior. Stopping a Host while it is in the middle of starting is similar to cancelling an operation. When we cancel an operation that is running, we throw an exception to signify that the operation hasn't been completed successfully. I think we should be doing the same here - throw an exception from `Start`, if the Host has been stopped while it was starting.

If an application wants to allow for stopping during start, it can catch the exception - just like it can catch an `OperationCanceledException`.

**Comment by maryamariyan** [comment]
Converted to draft PR since it won't be part of 6.0
